src/v/iceberg/schema_protobuf.cc

```diff
@@ -0,0 +1,188 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/schema_protobuf.h"
+
+#include "iceberg/conversion_outcome.h"
+#include "iceberg/datatypes.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <fmt/core.h>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <google/protobuf/descriptor.h>
+#include <google/protobuf/descriptor.pb.h>
+
+#include <deque>
+
+namespace iceberg {
+
+namespace {
+
+namespace pb = google::protobuf;
+
+/**
+ * Schema conversion from protocol buffers to internal Redpanda iceberg schema
+ */
+using field_outcome = conversion_outcome<nested_field_ptr>;
+using struct_outcome = conversion_outcome<struct_type>;
+using type_stack = std::deque<const pb::Descriptor*>;
+static constexpr int max_recursion_depth = 100;
+
+field_outcome
+from_protobuf(const pb::FieldDescriptor& fd, bool is_repeated, type_stack&);
+
+field_outcome success(const pb::FieldDescriptor& fd, field_type ft) {
+    return nested_field::create(
+      fd.number(), fd.name(), field_required::no, std::move(ft));
```

Comment:
It doesn't have to be in this PR, but just noting that at some point we're going to have to reconcile descriptors that contain multiple structs. When we transform such a schema, we'll need to deterministically assign each field a unique field id.

Longer term, I think this logic may live on the Iceberg coordinator, and workers will RPC out to get the Iceberg schema for a given Kafka schema id, especially when we support schema evolution